src/javascript/crypto/e2e/coname/client.js

```diff
@@ -0,0 +1,338 @@
+ * @return {e2e.coname.LookupResponse} The lookup response
+ */
+e2e.coname.Client.prototype.decodeLookupMessage_ = function(jsonString) {
+  var proto = this.proto;
+  var lookupProof = JSON.parse(jsonString);
+  var b64decode = goog.crypt.base64.decodeStringToByteArray;
+
+  // a lot of convertions before they can be verified
+  // TODO: except the use of encodeAB(), may be better to move them to server
+  lookupProof.index = b64decode(lookupProof.index);
+  lookupProof.index_proof = b64decode(lookupProof.index_proof);
+
+  lookupProof.tree_proof.neighbors = goog.array.map(
+      lookupProof.tree_proof.neighbors,
+      function(n) {return b64decode(n)});
+  lookupProof.tree_proof.existing_index = b64decode(
+      lookupProof.tree_proof.existing_index);
+  lookupProof.tree_proof.existing_entry_hash = b64decode(
+      lookupProof.tree_proof.existing_entry_hash);
+
+  goog.array.forEach(lookupProof.ratifications, function(r) {
+    var id, encoding = b64decode(r.head), rHH;
+
+    r.head = proto['TimestampedEpochHead'].decode(encoding);
+    rHH = r.head.head;
+    rHH.encoding = new Uint8Array(rHH.encodeAB());
+    rHH.previous_summary_hash = new Uint8Array(
+        rHH.previous_summary_hash.toBuffer());
+    rHH.root_hash = new Uint8Array(rHH.root_hash.toBuffer());
+    // drop nanos; seconds can be well represented by Number (max 2^53-1)
+    rHH.issue_time = rHH.issue_time.seconds.toNumber();
+    r.head.encoding = encoding;
+
+    for (id in r.signatures) {
+      r.signatures[id] = b64decode(r.signatures[id]);
+    }
+  });
+
+  var entry = b64decode(lookupProof.entry);
+  lookupProof.entry = proto['Entry'].decode(entry);
+  lookupProof.entry.profile_commitment = new Uint8Array(
+      lookupProof.entry.profile_commitment.toBuffer());
+  lookupProof.entry.encoding = entry;
+
+
+  var profile = b64decode(lookupProof.profile);
+  lookupProof.profile = proto['Profile'].decode(profile);
+  lookupProof.profile.encoding = profile;
+
+  return /** @type {e2e.coname.LookupResponse} */ (lookupProof);
+};
+
+
+/**
+ * Lookup and validate public keys for an email address
+ * @param {string} email The email address to look up a public key
+ * @param {function(string, Object)} callback The function to call with email
+ *  and pgp key as parameters
+ * @param {function(string, Error)} errback The function to call when the email
+ *  lacks proper keys
+ */
+e2e.coname.Client.prototype.lookup = function(email, callback, errback) {
+  // normalize the email address
+  email = email.toLowerCase();
+
+  var errback_ = function(e) { errback && errback(email, e); }.bind(this);
+  var realm = e2e.coname.getRealmByEmail(email);
+
+  if (realm === null) {
+    throw new e2e.error.InvalidArgumentsError(
+        'Lookup: no realm is found for ' + email);
+  }
+
+  this.initialize(function() {
+    var xhr = new XMLHttpRequest();
+    xhr.open('POST', realm.addr + '/lookup', true);
+    xhr.timeout = 5000; // 5s
```

Comment:
can we have a retry loop here with configurable number of retries, before we bail out?